### PR TITLE
Add serf/1.3.9 recipe

### PR DESCRIPTION
**Pull request by madebr**
Specify library name and version:  **serf/1.3.9**

Tested on gcc9@Linux

- [x] Depends on #814
- [x] Depends on #1462 
- [x] Depends on #1698

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/wiki#how-to-submit-a-pull-request) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

@fulara Please check this recipe.

### CODE REVIEW

**Comment by fulara** [comment]
did you patch the files somehow and forgot to include the patch?
WHen I am trying to test it I am getting an error:
print 'Warning: Used unknown variables:', ', '.join(unknown.keys())
line 186 of SConstruct file.
In the end for local recipe I just decided to take it from their master - where it is fixed.

Or is there something I am missing?

**Comment by madebr** [comment]
> did you patch the files somehow and forgot to include the patch?
> WHen I am trying to test it I am getting an error:

No patch.
That `SyntaxError` is emitted if you run scons without arguments.
When I pass some arguments, the build script just runs fine.
I haven't tested using clang/mingw/Visual Studio though.

**Comment by fulara** [comment]
running without arguments? Weird.. probably there is something borked with that SConstruct on the 1.3.9 and they only fixed it on master.. eh. 

this is the output I am getting,  long ass lines incoming!
```
 scons "-Y" "/var/build/fulara/conan-recipes/serf/all/conan-center-index/recipes/serf/all/build/source_subfolder" APR="/var/build/fulara/conan-standalone/.conan/data/apr/1.7.0/_/_/package/9fdb4217a0bb5bac441d7e17705c9172eeeb6cfe" APU="/var/build/fulara/conan-standalone/.conan/data/apr-util/1.6.1/_/_/package/89561afbdff4fa1d34b31088108f74fd0f951c38" OPENSSL="/var/build/fulara/conan-standalone/.conan/data/openssl/1.1.1f/_/_/package/5affb74e13c3238143a031c9fec992024696297b" PREFIX="/var/build/fulara/conan-recipes/serf/all/conan-center-index/recipes/serf/all/build/package" LIBDIR="/var/build/fulara/conan-recipes/serf/all/conan-center-index/recipes/serf/all/build/package/lib" ZLIB="/var/build/fulara/conan-standalone/.conan/data/zlib/1.2.11/_/_/package/9fdb4217a0bb5bac441d7e17705c9172eeeb6cfe" DEBUG="False" APR_STATIC="True" CFLAGS="-fPIC" LINKFLAGS=" -L'/var/build/fulara/conan-standalone/.conan/data/apr-util/1.6.1/_/_/package/89561afbdff4fa1d34b31088108f74fd0f951c38/lib' -L'/var/build/fulara/conan-standalone/.conan/data/apr-util/1.6.1/_/_/package/89561afbdff4fa1d34b31088108f74fd0f951c38/lib/apr-util-1' -L'/var/build/fulara/conan-standalone/.conan/data/python/3.6.10/_/_/package/87aeea7f85b30ca074b35fe00a9c03f106ae0e1c/lib' -L'/var/build/fulara/conan-standalone/.conan/data/apr/1.7.0/_/_/package/9fdb4217a0bb5bac441d7e17705c9172eeeb6cfe/lib' -L'/var/build/fulara/conan-standalone/.conan/data/openssl/1.1.1f/_/_/package/5affb74e13c3238143a031c9fec992024696297b/lib' -L'/var/build/fulara/conan-standalone/.conan/data/sqlite3/3.31.1/_/_/package/739b5053031f53097bee159873ba86c681d16e96/lib' -L'/var/build/fulara/conan-standalone/.conan/data/expat/2.2.9/_/_/package/9fdb4217a0bb5bac441d7e17705c9172eeeb6cfe/lib' -L'/var/build/fulara/conan-standalone/.conan/data/zlib/1.2.11/_/_/package/9fdb4217a0bb5bac441d7e17705c9172eeeb6cfe/lib'" CPPFLAGS="-DXML_STATIC -DAPR_DECLARE_STATIC -DAPU_DECLARE_STATIC -I'/var/build/fulara/conan-standalone/.conan/data/apr-util/1.6.1/_/_/package/89561afbdff4fa1d34b31088108f74fd0f951c38/include' -I'/var/build/fulara/conan-standalone/.conan/data/python/3.6.10/_/_/package/87aeea7f85b30ca074b35fe00a9c03f106ae0e1c/include' -I'/var/build/fulara/conan-standalone/.conan/data/apr/1.7.0/_/_/package/9fdb4217a0bb5bac441d7e17705c9172eeeb6cfe/include' -I'/var/build/fulara/conan-standalone/.conan/data/openssl/1.1.1f/_/_/package/5affb74e13c3238143a031c9fec992024696297b/include' -I'/var/build/fulara/conan-standalone/.conan/data/sqlite3/3.31.1/_/_/package/739b5053031f53097bee159873ba86c681d16e96/include' -I'/var/build/fulara/conan-standalone/.conan/data/expat/2.2.9/_/_/package/9fdb4217a0bb5bac441d7e17705c9172eeeb6cfe/include' -I'/var/build/fulara/conan-standalone/.conan/data/zlib/1.2.11/_/_/package/9fdb4217a0bb5bac441d7e17705c9172eeeb6cfe/include'"
-----------------
scons: Reading SConscript files ...
  File "/var/build/fulara/conan-recipes/serf/all/conan-center-index/recipes/serf/all/build/source_subfolder/SConstruct", line 186
```

**Comment by fulara** [comment]
When using master from github and then rest of your recipe it fails to compile. 

I feel actually bad for getting you involved in this. :( 
errors I am getting are ... 
/var/build/fulara/conan-recipes/serf/all/conan-center-index/recipes/serf/all/build/source_subfolder/buckets/ssl_buckets.c:345: error: dereferencing pointer to incomplete type

**Comment by madebr** [comment]
Using the current master, these are the arguments to gcc to compile that c source.
Maybe by finding a difference, you can find out what the problem is.

<details>
<summary>compiling ssl_bucket for static and shared library</summary>

static `ssl_bucket.c`:
```
gcc -o buckets/ssl_buckets.o -c -fPIC -std=c89 -Wdeclaration-after-statement -Wmissing-prototypes -Wall -O2 -pthread -DXML_STATIC -DAPR_DECLARE_STATIC -DAPU_DECLARE_STATIC -I'/home/maarten/.conan/data/apr-util/1.6.1/_/_/package/a229f57de15bff0238822a428947d3eebe0cd4e2/include' -I'/home/maarten/.conan/data/apr/1.7.0/_/_/package/6af9cc7cb931c5ad942174fd7838eb655717c709/include' -I'/home/maarten/.conan/data/libmysqlclient/8.0.17/_/_/package/bfe11e103319523e77cc827e33f3e1d8d2fa3a63/include' -I'/home/maarten/.conan/data/sqlite3/3.31.1/_/_/package/587735c139196dd3a993eb4148405475ee47f70d/include' -I'/home/maarten/.conan/data/expat/2.2.9/_/_/package/6af9cc7cb931c5ad942174fd7838eb655717c709/include' -I'/home/maarten/.conan/data/libpq/11.5/_/_/package/8e0939db49a1d312829524beb4d0b6824e47691d/include' -I'/home/maarten/.conan/data/openssl/1.1.1g/_/_/package/6af9cc7cb931c5ad942174fd7838eb655717c709/include' -I'/home/maarten/.conan/data/zlib/1.2.11/_/_/package/6af9cc7cb931c5ad942174fd7838eb655717c709/include' -DNDEBUG -DAPR_DECLARE_STATIC -DAPU_DECLARE_STATIC -DLINUX -D_REENTRANT -D_GNU_SOURCE -DSERF_HAVE_SSL_LOCKING_CALLBACKS -DSERF_HAVE_OPENSSL_MALLOC_INIT -DSERF_HAVE_OPENSSL_VERSION_NUM -DSERF_HAVE_OPENSSL_ALPN -DSERF_HAVE_OSSL_HANDSHAKE_STATE -DHAVE_STDBOOL_H -I. -I/home/maarten/.conan/data/serf/2.0.0.beta.20190905/_/_/build/df381a3e4cb94322a54f0a2b633fd9525bfaf85b/source_subfolder -I/include -I/include -I/home/maarten/.conan/data/zlib/1.2.11/_/_/package/6af9cc7cb931c5ad942174fd7838eb655717c709/include -I/home/maarten/.conan/data/openssl/1.1.1g/_/_/package/6af9cc7cb931c5ad942174fd7838eb655717c709/include /home/maarten/.conan/data/serf/2.0.0.beta.20190905/_/_/build/df381a3e4cb94322a54f0a2b633fd9525bfaf85b/source_subfolder/buckets/ssl_buckets.c
/home/maarten/.conan/data/serf/2.0.0.beta.20190905/_/_/build/df381a3e4cb94322a54f0a2b633fd9525bfaf85b/source_subfolder/buckets/ssl_buckets.c:1425:22: warning: 'ssl_id' defined but not used [-Wunused-function]
 1425 | static unsigned long ssl_id(void)
      |                      ^~~~~~
/home/maarten/.conan/data/serf/2.0.0.beta.20190905/_/_/build/df381a3e4cb94322a54f0a2b633fd9525bfaf85b/source_subfolder/buckets/ssl_buckets.c:1415:13: warning: 'ssl_lock' defined but not used [-Wunused-function]
 1415 | static void ssl_lock(int mode, int n, const char *file, int line)
      |             ^~~~~~~~
/home/maarten/.conan/data/serf/2.0.0.beta.20190905/_/_/build/df381a3e4cb94322a54f0a2b633fd9525bfaf85b/source_subfolder/buckets/ssl_buckets.c:1409:13: warning: 'ssl_dyn_destroy' defined but not used [-Wunused-function]
 1409 | static void ssl_dyn_destroy(CRYPTO_dynlock_value *l, const char *file,
      |             ^~~~~~~~~~~~~~~
/home/maarten/.conan/data/serf/2.0.0.beta.20190905/_/_/build/df381a3e4cb94322a54f0a2b633fd9525bfaf85b/source_subfolder/buckets/ssl_buckets.c:1398:13: warning: 'ssl_dyn_lock' defined but not used [-Wunused-function]
 1398 | static void ssl_dyn_lock(int mode, CRYPTO_dynlock_value *l, const char *file,
      |             ^~~~~~~~~~~~
/home/maarten/.conan/data/serf/2.0.0.beta.20190905/_/_/build/df381a3e4cb94322a54f0a2b633fd9525bfaf85b/source_subfolder/buckets/ssl_buckets.c:1385:30: warning: 'ssl_dyn_create' defined but not used [-Wunused-function]
 1385 | static CRYPTO_dynlock_value *ssl_dyn_create(const char* file, int line)
```


shared `ssl_bucket.c`:
```
gcc -o buckets/ssl_buckets.os -c -fPIC -std=c89 -Wdeclaration-after-statement -Wmissing-prototypes -Wall -O2 -pthread -fPIC -DXML_STATIC -DAPR_DECLARE_STATIC -DAPU_DECLARE_STATIC -I'/home/maarten/.conan/data/apr-util/1.6.1/_/_/package/a229f57de15bff0238822a428947d3eebe0cd4e2/include' -I'/home/maarten/.conan/data/apr/1.7.0/_/_/package/6af9cc7cb931c5ad942174fd7838eb655717c709/include' -I'/home/maarten/.conan/data/libmysqlclient/8.0.17/_/_/package/bfe11e103319523e77cc827e33f3e1d8d2fa3a63/include' -I'/home/maarten/.conan/data/sqlite3/3.31.1/_/_/package/587735c139196dd3a993eb4148405475ee47f70d/include' -I'/home/maarten/.conan/data/expat/2.2.9/_/_/package/6af9cc7cb931c5ad942174fd7838eb655717c709/include' -I'/home/maarten/.conan/data/libpq/11.5/_/_/package/8e0939db49a1d312829524beb4d0b6824e47691d/include' -I'/home/maarten/.conan/data/openssl/1.1.1g/_/_/package/6af9cc7cb931c5ad942174fd7838eb655717c709/include' -I'/home/maarten/.conan/data/zlib/1.2.11/_/_/package/6af9cc7cb931c5ad942174fd7838eb655717c709/include' -DNDEBUG -DAPR_DECLARE_STATIC -DAPU_DECLARE_STATIC -DLINUX -D_REENTRANT -D_GNU_SOURCE -DSERF_HAVE_SSL_LOCKING_CALLBACKS -DSERF_HAVE_OPENSSL_MALLOC_INIT -DSERF_HAVE_OPENSSL_VERSION_NUM -DSERF_HAVE_OPENSSL_ALPN -DSERF_HAVE_OSSL_HANDSHAKE_STATE -DHAVE_STDBOOL_H -I. -I/home/maarten/.conan/data/serf/2.0.0.beta.20190905/_/_/build/df381a3e4cb94322a54f0a2b633fd9525bfaf85b/source_subfolder -I/include -I/include -I/home/maarten/.conan/data/zlib/1.2.11/_/_/package/6af9cc7cb931c5ad942174fd7838eb655717c709/include -I/home/maarten/.conan/data/openssl/1.1.1g/_/_/package/6af9cc7cb931c5ad942174fd7838eb655717c709/include /home/maarten/.conan/data/serf/2.0.0.beta.20190905/_/_/build/df381a3e4cb94322a54f0a2b633fd9525bfaf85b/source_subfolder/buckets/ssl_buckets.c
/home/maarten/.conan/data/serf/2.0.0.beta.20190905/_/_/build/df381a3e4cb94322a54f0a2b633fd9525bfaf85b/source_subfolder/buckets/ssl_buckets.c:1425:22: warning: 'ssl_id' defined but not used [-Wunused-function]
 1425 | static unsigned long ssl_id(void)
      |                      ^~~~~~
/home/maarten/.conan/data/serf/2.0.0.beta.20190905/_/_/build/df381a3e4cb94322a54f0a2b633fd9525bfaf85b/source_subfolder/buckets/ssl_buckets.c:1415:13: warning: 'ssl_lock' defined but not used [-Wunused-function]
 1415 | static void ssl_lock(int mode, int n, const char *file, int line)
      |             ^~~~~~~~
/home/maarten/.conan/data/serf/2.0.0.beta.20190905/_/_/build/df381a3e4cb94322a54f0a2b633fd9525bfaf85b/source_subfolder/buckets/ssl_buckets.c:1409:13: warning: 'ssl_dyn_destroy' defined but not used [-Wunused-function]
 1409 | static void ssl_dyn_destroy(CRYPTO_dynlock_value *l, const char *file,
      |             ^~~~~~~~~~~~~~~
/home/maarten/.conan/data/serf/2.0.0.beta.20190905/_/_/build/df381a3e4cb94322a54f0a2b633fd9525bfaf85b/source_subfolder/buckets/ssl_buckets.c:1398:13: warning: 'ssl_dyn_lock' defined but not used [-Wunused-function]
 1398 | static void ssl_dyn_lock(int mode, CRYPTO_dynlock_value *l, const char *file,
      |             ^~~~~~~~~~~~
/home/maarten/.conan/data/serf/2.0.0.beta.20190905/_/_/build/df381a3e4cb94322a54f0a2b633fd9525bfaf85b/source_subfolder/buckets/ssl_buckets.c:1385:30: warning: 'ssl_dyn_create' defined but not used [-Wunused-function]
 1385 | static CRYPTO_dynlock_value *ssl_dyn_create(const char* file, int line)
```
</details>

**Comment by fulara** [comment]
hmm, weirdit seems to not like my ssl somehw, in my case it adds these defines:
```
-DSERF_NO_SSL_BIO_WRAPPERS -DSERF_NO_SSL_X509_STORE_WRAPPERS -DSERF_NO_SSL_X509_GET0_NOTBEFORE -DSERF_NO_SSL_X509_GET0_NOTAFTER -DSERF_NO_SSL_X509_GET0_CHAIN -DSERF_NO_SSL_ASN1_STRING_GET0_DATA
```
and your version has this  define:
`-DSERF_HAVE_OPENSSL_ALPN`
full line:
```
gcc -o buckets/ssl_buckets.o -c -fPIC -std=c89 -Wdeclaration-after-statement -Wmissing-prototypes -Wall -O2 -pthread -DXML_STATIC -DAPR_DECLARE_STATIC -DAPU_DECLARE_STATIC -I'/var/build/fulara/conan-standalone/.conan/data/apr-util/1.6.1/_/_/package/89561afbdff4fa1d34b31088108f74fd0f951c38/include' -I'/var/build/fulara/conan-standalone/.conan/data/python/3.6.10/_/_/package/87aeea7f85b30ca074b35fe00a9c03f106ae0e1c/include' -I'/var/build/fulara/conan-standalone/.conan/data/apr/1.7.0/_/_/package/9fdb4217a0bb5bac441d7e17705c9172eeeb6cfe/include' -I'/var/build/fulara/conan-standalone/.conan/data/openssl/1.1.1f/_/_/package/5affb74e13c3238143a031c9fec992024696297b/include' -I'/var/build/fulara/conan-standalone/.conan/data/sqlite3/3.31.1/_/_/package/739b5053031f53097bee159873ba86c681d16e96/include' -I'/var/build/fulara/conan-standalone/.conan/data/expat/2.2.9/_/_/package/9fdb4217a0bb5bac441d7e17705c9172eeeb6cfe/include' -I'/var/build/fulara/conan-standalone/.conan/data/zlib/1.2.11/_/_/package/9fdb4217a0bb5bac441d7e17705c9172eeeb6cfe/include' -DNDEBUG -DAPR_DECLARE_STATIC -DAPU_DECLARE_STATIC -DLINUX -D_REENTRANT -D_GNU_SOURCE -DSERF_NO_SSL_BIO_WRAPPERS -DSERF_NO_SSL_X509_STORE_WRAPPERS -DSERF_NO_SSL_X509_GET0_NOTBEFORE -DSERF_NO_SSL_X509_GET0_NOTAFTER -DSERF_NO_SSL_X509_GET0_CHAIN -DSERF_NO_SSL_ASN1_STRING_GET0_DATA -DSERF_HAVE_SSL_LOCKING_CALLBACKS -DSERF_HAVE_OPENSSL_MALLOC_INIT -DSERF_HAVE_OPENSSL_VERSION_NUM -DSERF_HAVE_OSSL_HANDSHAKE_STATE -DHAVE_STDBOOL_H -I. -I/var/build/fulara/conan-recipes/serf/all/conan-center-index/recipes/serf/all/t/source_subfolder -I/include -I/include -I/var/build/fulara/conan-standalone/.conan/data/zlib/1.2.11/_/_/package/9fdb4217a0bb5bac441d7e17705c9172eeeb6cfe/include -I/var/build/fulara/conan-standalone/.conan/data/openssl/1.1.1f/_/_/package/5affb74e13c3238143a031c9fec992024696297b/include /var/build/fulara/conan-recipes/serf/all/conan-center-index/recipes/serf/all/t/source_subfolder/buckets/ssl_buckets.c
``` checking further.

**Comment by madebr** [comment]
Check the first lines of the scons configure script:
```
scons: Reading SConscript files ...
Checking for GNU-compatible C compiler...yes
Checking for C header file openssl/applink.c... no
Checking for C function BIO_set_init()... yes
Checking for C function X509_STORE_get0_param()... yes
Checking for C function X509_get0_notBefore()... yes
Checking for C function X509_get0_notAfter()... yes
Checking for C function X509_STORE_CTX_get0_chain()... yes
Checking for C function ASN1_STRING_get0_data()... yes
Checking for C function CRYPTO_set_locking_callback()... yes
Checking for C function OPENSSL_malloc_init()... yes
Checking for C function SSL_library_init()... no
Checking for C function OpenSSL_version_num()... yes
Checking for C function SSL_set_alpn_protos()... yes
Checking for C type OSSL_HANDSHAKE_STATE... yes
Checking for C header file stdbool.h... yes
scons: done reading SConscript files.
scons: Building targets ...
```

Also see https://github.com/apache/serf/blob/1cb2ab783fb7182a7d5db64d428be30b6d779a90/SConstruct#L505-L530

I have just explicitly removed `openssl-devel` from my machine to be certain that my os packages did not mix with the conan packages.


> hmm, weirdit seems to not like my ssl somehw, in my case it adds these defines:
> 
> ```
> -DSERF_NO_SSL_BIO_WRAPPERS -DSERF_NO_SSL_X509_STORE_WRAPPERS -DSERF_NO_SSL_X509_GET0_NOTBEFORE -DSERF_NO_SSL_X509_GET0_NOTAFTER -DSERF_NO_SSL_X509_GET0_CHAIN -DSERF_NO_SSL_ASN1_STRING_GET0_DATA
> ```
> 
> and your version has this define:
> `-DSERF_HAVE_OPENSSL_ALPN`
> checking further..
> full line:
> 
> ```
> gcc -o buckets/ssl_buckets.o -c -fPIC -std=c89 -Wdeclaration-after-statement -Wmissing-prototypes -Wall -O2 -pthread -DXML_STATIC -DAPR_DECLARE_STATIC -DAPU_DECLARE_STATIC -I'/var/build/fulara/conan-standalone/.conan/data/apr-util/1.6.1/_/_/package/89561afbdff4fa1d34b31088108f74fd0f951c38/include' -I'/var/build/fulara/conan-standalone/.conan/data/python/3.6.10/_/_/package/87aeea7f85b30ca074b35fe00a9c03f106ae0e1c/include' -I'/var/build/fulara/conan-standalone/.conan/data/apr/1.7.0/_/_/package/9fdb4217a0bb5bac441d7e17705c9172eeeb6cfe/include' -I'/var/build/fulara/conan-standalone/.conan/data/openssl/1.1.1f/_/_/package/5affb74e13c3238143a031c9fec992024696297b/include' -I'/var/build/fulara/conan-standalone/.conan/data/sqlite3/3.31.1/_/_/package/739b5053031f53097bee159873ba86c681d16e96/include' -I'/var/build/fulara/conan-standalone/.conan/data/expat/2.2.9/_/_/package/9fdb4217a0bb5bac441d7e17705c9172eeeb6cfe/include' -I'/var/build/fulara/conan-standalone/.conan/data/zlib/1.2.11/_/_/package/9fdb4217a0bb5bac441d7e17705c9172eeeb6cfe/include' -DNDEBUG -DAPR_DECLARE_STATIC -DAPU_DECLARE_STATIC -DLINUX -D_REENTRANT -D_GNU_SOURCE -DSERF_NO_SSL_BIO_WRAPPERS -DSERF_NO_SSL_X509_STORE_WRAPPERS -DSERF_NO_SSL_X509_GET0_NOTBEFORE -DSERF_NO_SSL_X509_GET0_NOTAFTER -DSERF_NO_SSL_X509_GET0_CHAIN -DSERF_NO_SSL_ASN1_STRING_GET0_DATA -DSERF_HAVE_SSL_LOCKING_CALLBACKS -DSERF_HAVE_OPENSSL_MALLOC_INIT -DSERF_HAVE_OPENSSL_VERSION_NUM -DSERF_HAVE_OSSL_HANDSHAKE_STATE -DHAVE_STDBOOL_H -I. -I/var/build/fulara/conan-recipes/serf/all/conan-center-index/recipes/serf/all/t/source_subfolder -I/include -I/include -I/var/build/fulara/conan-standalone/.conan/data/zlib/1.2.11/_/_/package/9fdb4217a0bb5bac441d7e17705c9172eeeb6cfe/include -I/var/build/fulara/conan-standalone/.conan/data/openssl/1.1.1f/_/_/package/5affb74e13c3238143a031c9fec992024696297b/include /var/build/fulara/conan-recipes/serf/all/conan-center-index/recipes/serf/all/t/source_subfolder/buckets/ssl_buckets.c
> ```

**Comment by fulara** [comment]
yea, i was looking at that. thanks for posting yours. 
```
scons: Reading SConscript files ...
Checking for GNU-compatible C compiler...yes
Checking for C header file openssl/applink.c... no
Checking for C function BIO_set_init()... no
Checking for C function X509_STORE_get0_param()... no
Checking for C function X509_get0_notBefore()... no
Checking for C function X509_get0_notAfter()... no
Checking for C function X509_STORE_CTX_get0_chain()... no
Checking for C function ASN1_STRING_get0_data()... no
Checking for C function CRYPTO_set_locking_callback()... yes
Checking for C function OPENSSL_malloc_init()... yes
Checking for C function SSL_library_init()... no
Checking for C function OpenSSL_version_num()... yes
Checking for C function SSL_set_alpn_protos()... no
Checking for C type OSSL_HANDSHAKE_STATE... yes
Checking for C header file stdbool.h... yes
```

if not conf.CheckFunc('BIO_set_init', '#include <openssl/crypto.h>'):
  env.Append(CPPDEFINES=['SERF_NO_SSL_BIO_WRAPPERS'])

Now trying to understand how this magic CheckFunc works, but i have function `BIO_set_init` in the <bio.h>  not crypto.h

**Comment by fulara** [comment]
Okay, this Scons seems pretty .. bad?
This is the config.log:
```
.sconf_temp/conftest_2.c: In function 'main':
.sconf_temp/conftest_2.c:14: error: too few arguments to function 'BIO_set_init'
scons: Configure: no
```
So it fails because it didnt provide it correct number of arguments, where documentation states clearly that there is no, no argument option available.

**Comment by fulara** [comment]
I have checked this on clean install of rh7 ( completely different build machine ) - it works there.. 
And yes - there is no openssl-devel available there.
Looks like something is borked on that machine I am building, i'll get back to you.

**Comment by fulara** [comment]
on rh7 this builds. 
on rh6 the patches to make this work so far I have:
1) fix the print to add parenthesis and change ',' to '+' -> i have no idea why this is needed on rh6 but rh7, must have something to do with python but i am not sure.
2) change the line:
`env.File('serf.h').get_contents(),`
to:
`env.File('serf.h').get_contents().decode('utf-8'),`

3) still working out:  figure out why the PATH is overwritten, and the Scons uses '/usr/bin/ld' instead of my gccs ld.

**Comment by madebr** [comment]
@fulara 

I fixed 1 and 2 as you suggested.
It has something to do with differences between python 2 and python 3.

For point 3, I found out that the script always uses gcc, independent of what compiler you've set at the `CC` environment variable.
The script has a `CC` argument, so I used that.
Now it will also correctly build serf using clang on Linux.

Can you check whether this is enough to fix your build?\

Also, the current master of serf has a cmake script.
We will probably switch to that once a new version is released.

**Comment by fulara** [comment]
hello @madebr yes, I've been watching you magician doing your magic. 
I think you also need to pass the `path` somehow. or maybe it uses an LD variable or sth.
problem is that it takes a system ld (just tested) rather than one from toolset, and as I pasted above it breaks the build, ld crashes.

I thought that scons forces python 3+ but yea you are right, i think it forces python ~2.7 or 3.5+ 
and python 2 is fine with these shenanigans.

**Comment by fulara** [comment]
sorry @madebr  but this sitll does notw ork.
it still takes /usr/bin/ld with latest version.
actually isnt that the case that its the gcc that selects the LD? because it fails on gcc command.
Or maybe scons does not pass the LD to gcc?
```
gcc -o libserf-1.so.1.3.0 -L'/var/build/fulara/conan-standalone/.conan/data/apr-util/1.6.1/_/_/package/89561afbdff4fa1d34b31088108f74fd0f951c38/lib' -L'/var/build/fulara/conan-standalone/.conan/data/apr-util/1.6.1/_/_/package/89561afbdff4fa1d34b31088108f74fd0f951c38/lib/apr-util-1' -L'/var/build/fulara/conan-standalone/.conan/data/python/3.6.10/_/_/package/87aeea7f85b30ca074b35fe00a9c03f106ae0e1c/lib' -L'/var/build/fulara/conan-standalone/.conan/data/apr/1.7.0/_/_/package/9fdb4217a0bb5bac441d7e17705c9172eeeb6cfe/lib' -L'/var/build/fulara/conan-standalone/.conan/data/openssl/1.1.1f/_/_/package/5affb74e13c3238143a031c9fec992024696297b/lib' -L'/var/build/fulara/conan-standalone/.conan/data/sqlite3/3.31.1/_/_/package/739b5053031f53097bee159873ba86c681d16e96/lib' -L'/var/build/fulara/conan-standalone/.conan/data/expat/2.2.9/_/_/package/9fdb4217a0bb5bac441d7e17705c9172eeeb6cfe/lib' -L'/var/build/fulara/conan-standalone/.conan/data/zlib/1.2.11/_/_/package/9fdb4217a0bb5bac441d7e17705c9172eeeb6cfe/lib' -pthread -shared -Wl,-Bsymbolic -Wl,-soname=libserf-1.so.1 -Wl,-rpath=/var/build/fulara/conan-standalone/.conan/data/serf/1.3.9/_/_/package/1e90c5371aa64e39d79280e5e95b6611f302f0c5/lib -Wl,-rpath=/lib -Wl,-rpath=/var/build/fulara/conan-standalone/.conan/data/openssl/1.1.1f/_/_/package/5affb74e13c3238143a031c9fec992024696297b/lib context.os incoming.os outgoing.os ssltunnel.os buckets/aggregate_buckets.os buckets/allocator.os buckets/barrier_buckets.os buckets/buckets.os buckets/bwtp_buckets.os buckets/chunk_buckets.os buckets/dechunk_buckets.os buckets/deflate_buckets.os buckets/file_buckets.os buckets/headers_buckets.os buckets/iovec_buckets.os buckets/limit_buckets.os buckets/mmap_buckets.os buckets/request_buckets.os buckets/response_body_buckets.os buckets/response_buckets.os buckets/simple_buckets.os buckets/socket_buckets.os buckets/ssl_buckets.os auth/auth.os auth/auth_basic.os auth/auth_digest.os auth/auth_spnego.os auth/auth_spnego_gss.os auth/auth_spnego_sspi.os -L/lib -L/var/build/fulara/conan-standalone/.conan/data/openssl/1.1.1f/_/_/package/5affb74e13c3238143a031c9fec992024696297b/lib -lssl -lcrypto -lz -lapr-1 -lrt -lcrypt -lpthread -ldl -laprutil-1 -lexpat
/usr/bin/ld: BFD version 2.20.51.0.2-5.48.el6_10.1 20100205 internal error, aborting at reloc.c line 443 in bfd_get_reloc_size

/usr/bin/ld: Please report this bug.
```

I appreciate your patience, and apologies for pulling you into this.
speaking on pulling you through this - you are working on python conans.. by any chance areyou working on perl conans as well? :)

**Comment by madebr** [comment]
@fulara 

That error is an internal compiler error.
What happens if you chdir to `/var/build/fulara/conan-standalone/.conan/data/serf/1.3.9/_/_/build/1e90c5371aa64e39d79280e5e95b6611f302f0c5/` and run that `gcc` link command?
Maybe by reducing/modifying the arguments, you can find your problem.

**Comment by fulara** [comment]
@madebr  this is the error because the LD (or gcc as well didnt check (hard to check actually because .. serf ) is not taken from my path.
it should be using a system gcc bur a gcc provided from toolset ( which should be perfectly fine. )

when I invoke that same line from by hand it links fine ( because gcc and ld are taken from my path ).

**Comment by fulara** [comment]
@madebr build fine now, tthakns for your patience.

File: recipes/serf/all/conandata.yml [hunk 1]
```diff
@@ -0,0 +1,16 @@
+sources:
+  "1.3.9":
+    url: "https://www.apache.org/dist/serf/serf-1.3.9.tar.bz2"
```
[the remark this answers is not in the document]
Comment (madebr) [reply recipes/serf/all/conandata.yml:3]: I'll add it as an alternative.

File: recipes/serf/all/conandata.yml [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+    sha256: "549c2d21c577a8a9c0450facb5cca809f26591f048e466552240947bdf7a87cc"
+  "2.0.0.beta.20190905":
```
[the remark this answers is not in the document]
Comment (madebr) [reply recipes/serf/all/conandata.yml:5]: I've added 2.0.0 more as a poc for @fulara .
I'll it once `apr` and `apr-util` have made it on cci.

[the remark this answers is not in the document]
Comment (fulara) [reply recipes/serf/all/conandata.yml:5]: 1.3.9 works for me,  with your patch to Sconstruct - you can remove the 2.0.0. thanks. 💐

[the remark this answers is not in the document]
Comment (uilianries) [reply recipes/serf/all/conandata.yml]: nice!

File: recipes/serf/all/conandata.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (uilianries) [reply recipes/serf/all/conandata.yml:3]: I think you add it as mirror

**Comment by SSE4** [comment]
waiting on https://github.com/conan-io/conan-center-index/pull/1698 I guess?

**Comment by danimtb** [comment]
@madebr anything else to be done here after #1698 was merged?

**Comment by madebr** [comment]
@danimtb @uilianries @SSE4 
#1698 was merged, so I could revert the system_libs hack.
Plese re-review.

**Comment by madebr** [comment]
@SSE4 @uilianries @danimtb @jgsogo Reviews please 😄

**Comment by uilianries** [comment]
almost 2 weeks! :birthday:

File: recipes/serf/all/conandata.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (uilianries) [reply recipes/serf/all/conandata.yml:3]: @madebr did you add the mirror?

[the remark this answers is not in the document]
Comment (madebr) [reply recipes/serf/all/conandata.yml:3]: Nope, will do now.

File: recipes/serf/all/conandata.yml [hunk 3]
```diff
@@ -0,0 +1,12 @@
+sources:
+  "1.3.9":
+    url: "https://www.apache.org/dist/serf/serf-1.3.9.tar.bz2"
```
Comment (madebr) [inline recipes/serf/all/conandata.yml:3]:
```suggestion
    url: [
            "https://www.apache.org/dist/serf/serf-1.3.9.tar.bz2",
            "https://github.com/apache/serf/archive/1.3.9.tar.gz",
    ]
```

**Comment by madebr** [comment]
> 
> 
> almost 2 weeks! 🎂

2 months, you mean 🍰

**Comment by madebr** [comment]
@SSE4 @danimtb Please review!

**Pull request merged**